File: backend/app/authoring/service.py
"""Planning and atomic commit service shared by every Authoring import source."""
from __future__ import annotations

import logging
import os
import tempfile
from collections import Counter
from collections.abc import Callable
from pathlib import Path
from threading import Lock

import yaml
from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.authoring.schemas import (
    AuthoringImportIssue,
    AuthoringImportItem,
    AuthoringImportMode,
    AuthoringImportPreview,
    AuthoringImportResult,
    AuthoringImportSelection,
    ImportIssueSeverity,
    ImportItemStatus,
)
from app.authoring.sources import (
    ImportBundle,
    ImportResource,
    PlaylistPayload,
)
from app.models.playlist import Playlist, PlaylistItem
from app.models.track import Track
from app.modes import loader as modes_loader
from app.modes.loader import CueSpec, InterruptSpec, SoundboardManifest
from app.presets.loader import PresetManifest

logger = logging.getLogger(__name__)

# A commit spans SQLite and several mode files. Serializing this short critical
# section lets us re-plan immediately before writing and keeps imports
# create-only even when two requests race.
_import_lock = Lock()


def mode_or_404(mode_id: str) -> modes_loader.ModeManifest:
    mode = modes_loader.get_mode(mode_id)
    if mode is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"mode '{mode_id}' not loaded",
        )
    return mode


def _selection_key(selection: AuthoringImportSelection) -> str:
    return f"{selection.kind}:{selection.resource_id}"


def _issue(
    code: str,
    severity: ImportIssueSeverity,
    message: str,
    *,
    related: ImportResource | None = None,
) -> AuthoringImportIssue:
    return AuthoringImportIssue(
        code=code,
        severity=severity,
        message=message,
        related_item=(
            AuthoringImportSelection(
                kind=related.kind,
                resource_id=related.resource_id,
            )
            if related
            else None
        ),
    )


def _soundboard_contains(soundboard: SoundboardManifest, item_path: str) -> bool:
    return any(
        item.file == item_path
        for category in soundboard.categories
        for item in category.items
    )


def _dependency_issues(
    resource: ImportResource,
    bundle: ImportBundle,
    target: modes_loader.ModeManifest,
    target_playlist_names: set[str],
) -> list[AuthoringImportIssue]:
    """Describe references that require another selection or cannot resolve."""

    source_playlists: dict[str, list[ImportResource]] = {}
    source_presets: dict[str, ImportResource] = {}
    source_soundboards: dict[str, ImportResource] = {}
    for candidate in bundle.resources:
        if candidate.kind == "playlist" and isinstance(
            candidate.payload, PlaylistPayload
        ):
            source_playlists.setdefault(candidate.payload.name, []).append(candidate)
        elif candidate.kind == "preset":
            source_presets[candidate.resource_id] = candidate
        elif candidate.kind == "soundboard":
            source_soundboards[candidate.resource_id] = candidate

    issues: list[AuthoringImportIssue] = []

    def require_playlist(name: str) -> None:
        if name in target_playlist_names:
            return
        candidates = source_playlists.get(name, [])
        if len(candidates) == 1:
            issues.append(
                _issue(
                    "dependency_selection_required",
                    "warning",
                    f"Also select playlist '{name}'.",
                    related=candidates[0],
                )
            )
        elif len(candidates) > 1:
            issues.append(
                _issue(
                    "ambiguous_dependency",
                    "error",
                    f"Playlist reference '{name}' matches multiple source playlists.",
                )
            )
        else:
            issues.append(
                _issue(
                    "missing_dependency",
                    "error",
                    f"Referenced playlist '{name}' is not in the target or import document.",
                )
            )

    def require_preset(preset_id: str) -> None:
        if preset_id in target.presets:
            return
        candidate = source_presets.get(preset_id)
        if candidate:
            issues.append(
                _issue(
                    "dependency_selection_required",
                    "warning",
                    f"Also select EQ preset '{preset_id}'.",
                    related=candidate,
                )
            )
        else:
            issues.append(
                _issue(
                    "missing_dependency",
                    "error",
                    f"Referenced EQ preset '{preset_id}' is not in the target or import document.",
                )
            )

    def require_soundboard(soundboard_id: str, item_path: str) -> None:
        target_board = target.soundboards.get(soundboard_id)
        if target_board:
            if _soundboard_contains(target_board, item_path):
                return
            issues.append(
                _issue(
                    "missing_dependency",
                    "error",
                    (
                        f"Target soundboard '{soundboard_id}' does not contain "
                        f"sound '{item_path}', and its ID is already occupied."
                    ),
                )
            )
            return
        candidate = source_soundboards.get(soundboard_id)
        if (
            candidate
            and isinstance(candidate.payload, SoundboardManifest)
            and _soundboard_contains(candidate.payload, item_path)
        ):
            issues.append(
                _issue(
                    "dependency_selection_required",
                    "warning",
                    f"Also select soundboard '{soundboard_id}'.",
                    related=candidate,
                )
            )
            return
        issues.append(
            _issue(
                "missing_dependency",
                "error",
                f"Sound '{item_path}' is not available in soundboard '{soundboard_id}'.",
            )
        )

    def require_sound_path(item_path: str) -> None:
        if any(
            _soundboard_contains(soundboard, item_path)
            for soundboard in target.soundboards.values()
        ):
            return
        matches = [
            candidate
            for candidate in source_soundboards.values()
            if isinstance(candidate.payload, SoundboardManifest)
            and _soundboard_contains(candidate.payload, item_path)
        ]
        if len(matches) == 1:
            if matches[0].resource_id in target.soundboards:
                issues.append(
                    _issue(
                        "missing_dependency",
                        "error",
                        (
                            f"Target soundboard '{matches[0].resource_id}' does not "
                            f"contain sound '{item_path}', and its ID is already occupied."
                        ),
                    )
                )
                return
            issues.append(
                _issue(
                    "dependency_selection_required",
                    "warning",
                    f"Also select soundboard '{matches[0].resource_id}'.",
                    related=matches[0],
                )
            )
        elif len(matches) > 1:
            issues.append(
                _issue(
                    "ambiguous_dependency",
                    "error",
                    f"Sound reference '{item_path}' matches multiple source soundboards.",
                )
            )
        else:
            issues.append(
                _issue(
                    "missing_dependency",
                    "error",
                    f"Referenced sound '{item_path}' is not in the target or import document.",
                )
            )

    if isinstance(resource.payload, InterruptSpec):
        if resource.payload.playlist:
            require_playlist(resource.payload.playlist)
        elif resource.payload.soundboard_item:
            require_sound_path(resource.payload.soundboard_item)
    elif isinstance(resource.payload, CueSpec):
        if resource.payload.preset:
            require_preset(resource.payload.preset)
        if resource.payload.playlist:
            require_playlist(resource.payload.playlist)
        for sfx in resource.payload.sfx:
            require_soundboard(sfx.soundboard, sfx.item)
        for loop in resource.payload.loops:
            require_soundboard(loop.soundboard, loop.item)

    return issues


def build_preview(
    db: Session,
    target_mode_id: str,
    bundle: ImportBundle,
) -> AuthoringImportPreview:
    target = mode_or_404(target_mode_id)
    if bundle.source.type == "mode" and bundle.source.id == target_mode_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="source and target modes must be different",
        )

    target_playlist_names = set(
        db.scalars(
            select(Playlist.name).where(Playlist.mode_id == target_mode_id)
        ).all()
    )
    playlist_name_counts = Counter(
        resource.payload.name
        for resource in bundle.resources
        if resource.kind == "playlist"
        and isinstance(resource.payload, PlaylistPayload)
    )
    interrupt_name_counts = Counter(
        resource.name
        for resource in bundle.resources
        if resource.kind == "interrupt"
    )
    target_interrupt_names = {interrupt.name for interrupt in target.interrupts}
    library_paths = set(db.scalars(select(Track.path)).all())

    items: list[AuthoringImportItem] = []
    for resource in bundle.resources:
        issues = list(resource.issues)
        conflict_reason: str | None = None

        if resource.kind == "playlist" and isinstance(
            resource.payload, PlaylistPayload
        ):
            if playlist_name_counts[resource.payload.name] > 1:
                issues.append(
                    _issue(
                        "duplicate_source_name",
                        "error",
                        "Another source playlist has the same name.",
                    )
                )
            elif resource.payload.name in target_playlist_names:
                conflict_reason = (
                    "A playlist with this name already exists in the target mode."
                )
            missing = [
                track.missing_label
                for track in resource.payload.tracks
                if track.path is None or track.path not in library_paths
            ]
            if missing:
                issues.append(
                    _issue(
                        "missing_tracks",
                        "warning",
                        f"{len(missing)} track reference(s) are unavailable and will be omitted.",
                    )
                )
        elif resource.kind == "soundboard":
            if resource.resource_id in target.soundboards:
                conflict_reason = (
                    "A soundboard with this ID already exists in the target mode."
                )
        elif resource.kind == "interrupt":
            if interrupt_name_counts[resource.name] > 1:
                issues.append(
                    _issue(
                        "duplicate_source_name",
                        "error",
                        "Another source interrupt has the same name.",
                    )
                )
            elif resource.name in target_interrupt_names:
                conflict_reason = (
                    "An interrupt with this name already exists in the target mode."
                )
        elif resource.kind == "preset":
            if resource.resource_id in target.presets:
                conflict_reason = (
                    "An EQ preset with this ID already exists in the target mode."
                )
        elif resource.resource_id in target.cues:
            conflict_reason = "A cue with this ID already exists in the target mode."

        issues.extend(
            _dependency_issues(
                resource,
                bundle,
                target,
                target_playlist_names,
            )
        )
        first_error = next(
            (issue.message for issue in issues if issue.severity == "error"),
            None,
        )
        if conflict_reason:
            item_status: ImportItemStatus = "conflict"
            reason = conflict_reason
            issues.append(_issue("target_conflict", "error", conflict_reason))
        elif first_error:
            item_status = "invalid"
            reason = first_error
        else:
            item_status = "ready"
            reason = None

        items.append(
            AuthoringImportItem(
                kind=resource.kind,
                resource_id=resource.resource_id,
                name=resource.name,
                summary=resource.summary,
                status=item_status,
                reason=reason,
                issues=issues,
            )
        )

    return AuthoringImportPreview(
        source=bundle.source,
        source_mode=(
            AuthoringImportMode(id=bundle.source.id, name=bundle.source.name)
            if bundle.source.type == "mode"
            else None
        ),
        target_mode=AuthoringImportMode(id=target.id, name=target.name),
        items=items,
    )


def _write_yaml(path: Path, payload: dict) -> None:
    """Atomically create or replace one validated Authoring YAML document."""

    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temp_name = tempfile.mkstemp(
        dir=path.parent,
        prefix=f".{path.name}.",
        suffix=".tmp",
        text=True,
    )
    temp_path = Path(temp_name)
    try:
        with os.fdopen(fd, "w", encoding="utf-8", newline="\n") as temp_file:
            yaml.safe_dump(payload, temp_file, sort_keys=False)
            temp_file.flush()
            os.fsync(temp_file.fileno())
        os.replace(temp_path, path)
    except BaseException:
        temp_path.unlink(missing_ok=True)
        raise


def _write_bytes(path: Path, payload: bytes) -> None:
    """Atomically restore exact pre-import bytes during rollback."""

    fd, temp_name = tempfile.mkstemp(
        dir=path.parent,
        prefix=f".{path.name}.",
        suffix=".rollback",
        text=False,
    )
    temp_path = Path(temp_name)
    try:
        with os.fdopen(fd, "wb") as temp_file:
            temp_file.write(payload)
            temp_file.flush()
            os.fsync(temp_file.fileno())
        os.replace(temp_path, path)
    except BaseException:
        temp_path.unlink(missing_ok=True)
        raise


def _target_manifest(
    mode: modes_loader.ModeManifest,
) -> tuple[Path, dict, bytes]:
    if mode.root_dir is None:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"mode '{mode.id}' has no resolved root directory",
        )
    path = mode.root_dir / "manifest.yaml"
    original = path.read_bytes()
    raw = yaml.safe_load(original.decode("utf-8")) or {}
    if not isinstance(raw, dict):
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"manifest.yaml for mode '{mode.id}' is not a mapping",
        )
    for field_name in ("playlist_categories", "interrupts"):
        raw.setdefault(field_name, [])
        if not isinstance(raw[field_name], list):
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=(
                    f"manifest.yaml field '{field_name}' for mode "
                    f"'{mode.id}' is not a list"
                ),
            )
    return (path, raw, original)


def _selected_resources(
    preview: AuthoringImportPreview,
    bundle: ImportBundle,
    selections: list[AuthoringImportSelection],
) -> tuple[list[ImportResource], list[AuthoringImportItem]]:
    preview_by_key = {
        _selection_key(
            AuthoringImportSelection(kind=item.kind, resource_id=item.resource_id)
        ): item
        for item in preview.items
    }
    resources_by_key = bundle.by_key()
    requested_keys: list[str] = []
    seen: set[str] = set()
    for selection in selections:
        key = _selection_key(selection)
        if key in seen:
            continue
        if key not in preview_by_key or key not in resources_by_key:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"source resource '{key}' is no longer available",
            )
        requested_keys.append(key)
        seen.add(key)

    selected_ready_keys = {
        key for key in requested_keys if preview_by_key[key].status == "ready"
    }
    for key in selected_ready_keys:
        item = preview_by_key[key]
        for issue in item.issues:
            if issue.code != "dependency_selection_required" or not issue.related_item:
                continue
            dependency_key = _selection_key(issue.related_item)
            if dependency_key not in selected_ready_keys:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=(
                        f"'{item.name}' requires {issue.related_item.kind} "
                        f"'{issue.related_item.resource_id}' to be selected and ready"
                    ),
                )

    imported = [resources_by_key[key] for key in requested_keys if key in selected_ready_keys]
    skipped = [
        preview_by_key[key] for key in requested_keys if key not in selected_ready_keys
    ]
    return imported, skipped


def commit_bundle(
    db: Session,
    target_mode_id: str,
    bundle_factory: Callable[[], ImportBundle],
    selections: list[AuthoringImportSelection],
) -> AuthoringImportResult:
    """Re-plan and atomically create a selection in the target mode."""

    with _import_lock:
        bundle = bundle_factory()
        preview = build_preview(db, target_mode_id, bundle)
        imported_resources, skipped = _selected_resources(
            preview,
            bundle,
            selections,
        )
        if not imported_resources:
            return AuthoringImportResult(
                imported=[],
                skipped=skipped,
                missing_track_paths=[],
            )

        target = mode_or_404(target_mode_id)
        if target.root_dir is None:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"mode '{target.id}' has no resolved root directory",
            )

        manifest_path, manifest, original_manifest = _target_manifest(target)
        manifest_changed = False
        file_payloads: list[tuple[Path, dict]] = []
        missing_track_paths: list[str] = []

        library_tracks = {
            track.path: track
            for track in db.scalars(select(Track).order_by(Track.id)).all()
        }
        for resource in imported_resources:
            if isinstance(resource.payload, PlaylistPayload):
                cloned = Playlist(
                    name=resource.payload.name,
                    mode_id=target.id,
                    category=resource.payload.category,
                )
                db.add(cloned)
                db.flush()
                next_position = 0
                for track_ref in resource.payload.tracks:
                    track = (
                        library_tracks.get(track_ref.path)
                        if track_ref.path is not None
                        else None
                    )
                    if track is None:
                        missing_track_paths.append(track_ref.missing_label)
                        continue
                    db.add(
                        PlaylistItem(
                            playlist_id=cloned.id,
                            position=next_position,
                            track_id=track.id,
                        )
                    )
                    next_position += 1
                category = resource.payload.category
                if category and category not in manifest["playlist_categories"]:
                    manifest["playlist_categories"].append(category)
                    manifest_changed = True
            elif isinstance(resource.payload, SoundboardManifest):
                file_payloads.append(
                    (
                        target.root_dir
                        / "soundboards"
                        / f"{resource.resource_id}.yaml",
                        resource.payload.model_dump(exclude_none=True),
                    )
                )
            elif isinstance(resource.payload, InterruptSpec):
                manifest["interrupts"].append(
                    resource.payload.model_dump(exclude_none=True)
                )
                manifest_changed = True
            elif isinstance(resource.payload, PresetManifest):
                file_payloads.append(
                    (
                        target.root_dir / "presets" / f"{resource.resource_id}.yaml",
                        resource.payload.model_dump(exclude_none=True),
                    )
                )
            elif isinstance(resource.payload, CueSpec):
                file_payloads.append(
                    (
                        target.root_dir / "cues" / f"{resource.resource_id}.yaml",
                        resource.payload.model_dump(exclude_none=True),
                    )
                )

        created_files: list[Path] = []
        wrote_manifest = False
        try:
            db.flush()
            for output_path, document in file_payloads:
                # Other Authoring endpoints do not share this lock. A final
                # existence check preserves the create-only contract.
                if output_path.exists():
                    raise HTTPException(
                        status_code=status.HTTP_409_CONFLICT,
                        detail=(
                            "target resource appeared during import: "
                            f"{output_path.name}"
                        ),
                    )
                _write_yaml(output_path, document)
                created_files.append(output_path)
            if manifest_changed:
                _write_yaml(manifest_path, manifest)
                wrote_manifest = True

            modes_loader.reload_mode(target.id)
            db.commit()
        except Exception as exc:
            db.rollback()
            rollback_errors: list[str] = []
            for created_path in reversed(created_files):
                try:
                    created_path.unlink(missing_ok=True)
                except OSError as rollback_error:
                    rollback_errors.append(
                        f"could not remove {created_path.name}: {rollback_error}"
                    )
            if wrote_manifest:
                try:
                    _write_bytes(manifest_path, original_manifest)
                except OSError as rollback_error:
                    rollback_errors.append(
                        f"could not restore manifest.yaml: {rollback_error}"
                    )
            try:
                modes_loader.reload_mode(target.id)
            except Exception as rollback_error:
                logger.exception("failed to reload target mode after import rollback")
                rollback_errors.append(
                    f"could not reload target mode: {rollback_error}"
                )
            if rollback_errors:
                logger.error(
                    "authoring import rollback incomplete: %s",
                    "; ".join(rollback_errors),
                )
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=(
                        "authoring import failed and rollback was incomplete; "
                        "check server logs before retrying"
                    ),
                ) from exc
            if isinstance(exc, HTTPException):
                raise
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=(
                    "authoring import failed and was rolled back: "
                    f"{type(exc).__name__}: {exc}"
                ),
            ) from exc

        imported_items_by_key = {
            f"{item.kind}:{item.resource_id}": item for item in preview.items
        }
        return AuthoringImportResult(
            imported=[
                imported_items_by_key[resource.key]
                for resource in imported_resources
            ],
            skipped=skipped,
            missing_track_paths=missing_track_paths,
        )

File: crates/music-server/src/authoring/service.rs
use std::collections::{BTreeMap, BTreeSet};

use music_application::modes::{
    CueDocument, CueLoopDocument, CueSfxDocument, EffectDocument, InterruptDocument, ModeBundle,
    ModeCatalog, ModeDocument, ModeImportPlaylist, ModeMutation, ModeMutationError,
    ModeMutationFailureKind, PresetDocument, SoundboardCategoryDocument, SoundboardDocument,
    SoundboardItemDocument,
};
use music_application::playlists::PlaylistFilter;
use music_domain::{LibraryPath, TrackId};

use super::model::{
    AuthoringImportDocumentV1, AuthoringImportIssue, AuthoringImportItem, AuthoringImportMode,
    AuthoringImportPreview, AuthoringImportResult, AuthoringImportSelection, AuthoringImportSource,
    AuthoringResourceKind, AuthoringSourceType, ImportIssueSeverity, ImportItemStatus,
};
use crate::error::ApiError;
use crate::http::HttpState;

mod dependencies;
mod planning;
mod source;

use dependencies::dependency_issues;
use planning::{build_mutation, select_resources};
use source::load_bundle;

const MAX_COMMIT_REPLANS: usize = 3;

#[derive(Debug, Clone)]
pub(super) enum ImportSourceSpec {
    Mode(String),
    Document {
        document: AuthoringImportDocumentV1,
        source_name: Option<String>,
    },
}

#[derive(Debug, Clone)]
struct PlaylistTrackRef {
    path: Option<String>,
    missing_label: String,
}

#[derive(Debug, Clone)]
struct PlaylistPayload {
    name: String,
    category: Option<String>,
    tracks: Vec<PlaylistTrackRef>,
}

#[derive(Debug, Clone)]
enum ResourcePayload {
    Playlist(PlaylistPayload),
    Soundboard(SoundboardDocument),
    Interrupt(InterruptDocument),
    Preset(PresetDocument),
    Cue(CueDocument),
}

#[derive(Debug, Clone)]
struct ImportResource {
    kind: AuthoringResourceKind,
    resource_id: String,
    name: String,
    summary: String,
    payload: ResourcePayload,
    issues: Vec<AuthoringImportIssue>,
}

impl ImportResource {
    fn key(&self) -> String {
        format!("{}:{}", self.kind.as_str(), self.resource_id)
    }

    fn selection(&self) -> AuthoringImportSelection {
        AuthoringImportSelection {
            kind: self.kind,
            resource_id: self.resource_id.clone(),
        }
    }
}

#[derive(Debug, Clone)]
struct ImportBundle {
    source: AuthoringImportSource,
    resources: Vec<ImportResource>,
}

struct PreviewPlan {
    preview: AuthoringImportPreview,
    target: ModeBundle,
    library_tracks: BTreeMap<String, TrackId>,
}

struct SelectionPlan {
    imported: Vec<ImportResource>,
    skipped: Vec<AuthoringImportItem>,
}

struct MutationPlan {
    mutation: ModeMutation,
    result: AuthoringImportResult,
}

pub(super) async fn preview(
    state: &HttpState,
    target_mode_id: &str,
    source: &ImportSourceSpec,
) -> Result<AuthoringImportPreview, ApiError> {
    let catalog = mode_catalog(state)?;
    let bundle = load_bundle(state, source, &catalog).await?;
    Ok(build_preview(state, target_mode_id, bundle, &catalog)
        .await?
        .preview)
}

pub(super) async fn commit(
    state: &HttpState,
    target_mode_id: &str,
    source: &ImportSourceSpec,
    selections: &[AuthoringImportSelection],
) -> Result<AuthoringImportResult, ApiError> {
    let mut last_conflict = None;
    for _ in 0..MAX_COMMIT_REPLANS {
        let catalog = mode_catalog(state)?;
        let bundle = load_bundle(state, source, &catalog).await?;
        let plan = build_preview(state, target_mode_id, bundle.clone(), &catalog).await?;
        let selected = select_resources(&plan.preview, &bundle, selections)?;
        if selected.imported.is_empty() {
            return Ok(AuthoringImportResult {
                imported: Vec::new(),
                skipped: selected.skipped,
                missing_track_paths: Vec::new(),
            });
        }
        let mutation = build_mutation(catalog.generation, target_mode_id, plan, selected);
        let coordinator = state
            .modes
            .as_ref()
            .ok_or_else(ApiError::service_unavailable)?;
        match coordinator.mutate(mutation.mutation).await {
            Ok(_) => return Ok(mutation.result),
            Err(error)
                if matches!(
                    error.kind,
                    ModeMutationFailureKind::Stale | ModeMutationFailureKind::Conflict
                ) =>
            {
                last_conflict = Some(error);
            }
            Err(error) => return Err(map_mutation_error(error)),
        }
    }
    Err(last_conflict.map_or_else(
        || ApiError::conflict("authoring import changed during commit"),
        map_mutation_error,
    ))
}

fn mode_catalog(state: &HttpState) -> Result<std::sync::Arc<ModeCatalog>, ApiError> {
    state
        .modes
        .as_ref()
        .and_then(music_application::modes::ModeCoordinatorHandle::snapshot)
        .ok_or_else(ApiError::service_unavailable)
}

async fn build_preview(
    state: &HttpState,
    target_mode_id: &str,
    bundle: ImportBundle,
    catalog: &ModeCatalog,
) -> Result<PreviewPlan, ApiError> {
    let target = catalog.modes.get(target_mode_id).cloned().ok_or_else(|| {
        ApiError::not_found_message(format!("mode '{target_mode_id}' not loaded"))
    })?;
    if bundle.source.source_type == AuthoringSourceType::Mode && bundle.source.id == target_mode_id
    {
        return Err(ApiError::bad_request(
            "source and target modes must be different",
        ));
    }
    let playlists = state
        .playlists
        .as_ref()
        .ok_or_else(ApiError::service_unavailable)?;
    let target_playlist_names = playlists
        .list(&PlaylistFilter {
            mode_id: Some(target_mode_id.to_owned()),
            category: None,
        })
        .await
        .map_err(|_| ApiError::service_unavailable())?
        .into_iter()
        .map(|playlist| playlist.name)
        .collect::<BTreeSet<_>>();
    let library = state
        .library
        .as_ref()
        .ok_or_else(ApiError::service_unavailable)?;
    let library_tracks = library
        .service
        .all_tracks()
        .await
        .map_err(|_| ApiError::service_unavailable())?
        .into_iter()
        .map(|track| (track.path.into_string(), track.id))
        .collect::<BTreeMap<_, _>>();

    let playlist_name_counts = counts(bundle.resources.iter().filter_map(|resource| {
        if let ResourcePayload::Playlist(playlist) = &resource.payload {
            Some(playlist.name.as_str())
        } else {
            None
        }
    }));
    let interrupt_name_counts = counts(bundle.resources.iter().filter_map(|resource| {
        matches!(resource.payload, ResourcePayload::Interrupt(_)).then_some(resource.name.as_str())
    }));
    let target_interrupt_names = target
        .manifest
        .interrupts
        .iter()
        .map(|interrupt| interrupt.name.as_str())
        .collect::<BTreeSet<_>>();

    let mut items = Vec::with_capacity(bundle.resources.len());
    for resource in &bundle.resources {
        let mut issues = resource.issues.clone();
        let mut conflict_reason = None;
        match &resource.payload {
            ResourcePayload::Playlist(playlist) => {
                if playlist_name_counts
                    .get(&playlist.name)
                    .copied()
                    .unwrap_or(0)
                    > 1
                {
                    issues.push(issue(
                        "duplicate_source_name",
                        ImportIssueSeverity::Error,
                        "Another source playlist has the same name.",
                        None,
                    ));
                } else if target_playlist_names.contains(&playlist.name) {
                    conflict_reason = Some(
                        "A playlist with this name already exists in the target mode.".to_owned(),
                    );
                }
                let missing = playlist
                    .tracks
                    .iter()
                    .filter(|track| {
                        track
                            .path
                            .as_ref()
                            .is_none_or(|path| !library_tracks.contains_key(path))
                    })
                    .count();
                if missing > 0 {
                    issues.push(issue(
                        "missing_tracks",
                        ImportIssueSeverity::Warning,
                        format!(
                            "{missing} track reference(s) are unavailable and will be omitted."
                        ),
                        None,
                    ));
                }
            }
            ResourcePayload::Soundboard(_) => {
                if target.soundboards.contains_key(&resource.resource_id) {
                    conflict_reason = Some(
                        "A soundboard with this ID already exists in the target mode.".to_owned(),
                    );
                }
            }
            ResourcePayload::Interrupt(_) => {
                if interrupt_name_counts
                    .get(&resource.name)
                    .copied()
                    .unwrap_or(0)
                    > 1
                {
                    issues.push(issue(
                        "duplicate_source_name",
                        ImportIssueSeverity::Error,
                        "Another source interrupt has the same name.",
                        None,
                    ));
                } else if target_interrupt_names.contains(resource.name.as_str()) {
                    conflict_reason = Some(
                        "An interrupt with this name already exists in the target mode.".to_owned(),
                    );
                }
            }
            ResourcePayload::Preset(_) => {
                if target.presets.contains_key(&resource.resource_id) {
                    conflict_reason = Some(
                        "An EQ preset with this ID already exists in the target mode.".to_owned(),
                    );
                }
            }
            ResourcePayload::Cue(_) => {
                if target.cues.contains_key(&resource.resource_id) {
                    conflict_reason =
                        Some("A cue with this ID already exists in the target mode.".to_owned());
                }
            }
        }
        issues.extend(dependency_issues(
            resource,
            &bundle,
            &target,
            &target_playlist_names,
        ));
        let first_error = issues
            .iter()
            .find(|issue| issue.severity == ImportIssueSeverity::Error)
            .map(|issue| issue.message.clone());
        let (status, reason) = if let Some(conflict) = conflict_reason {
            issues.push(issue(
                "target_conflict",
                ImportIssueSeverity::Error,
                conflict.clone(),
                None,
            ));
            (ImportItemStatus::Conflict, Some(conflict))
        } else if let Some(error) = first_error {
            (ImportItemStatus::Invalid, Some(error))
        } else {
            (ImportItemStatus::Ready, None)
        };
        items.push(AuthoringImportItem {
            kind: resource.kind,
            resource_id: resource.resource_id.clone(),
            name: resource.name.clone(),
            summary: resource.summary.clone(),
            status,
            reason,
            issues,
        });
    }

    let source_mode =
        (bundle.source.source_type == AuthoringSourceType::Mode).then(|| AuthoringImportMode {
            id: bundle.source.id.clone(),
            name: bundle.source.name.clone(),
        });
    Ok(PreviewPlan {
        preview: AuthoringImportPreview {
            source: bundle.source,
            source_mode,
            target_mode: AuthoringImportMode {
                id: target.manifest.id.clone(),
                name: target.manifest.name.clone(),
            },
            items,
        },
        target,
        library_tracks,
    })
}

fn issue(
    code: impl Into<String>,
    severity: ImportIssueSeverity,
    message: impl Into<String>,
    related: Option<&ImportResource>,
) -> AuthoringImportIssue {
    AuthoringImportIssue {
        code: code.into(),
        severity,
        message: message.into(),
        related_item: related.map(ImportResource::selection),
    }
}

fn counts<'a>(values: impl Iterator<Item = &'a str>) -> BTreeMap<String, usize> {
    let mut counts = BTreeMap::new();
    for value in values {
        *counts.entry(value.to_owned()).or_insert(0) += 1;
    }
    counts
}

fn map_mutation_error(error: ModeMutationError) -> ApiError {
    match error.kind {
        ModeMutationFailureKind::Invalid => ApiError::bad_request(error.code),
        ModeMutationFailureKind::NotFound => ApiError::plain_not_found(error.code),
        ModeMutationFailureKind::Conflict | ModeMutationFailureKind::Stale => {
            ApiError::conflict(error.code)
        }
        ModeMutationFailureKind::Unavailable => {
            tracing::error!(error = %error, "authoring import mutation failed");
            ApiError::service_unavailable()
        }
    }
}
